sandbox/pc2sandbox.sh

```diff
@@ -0,0 +1,234 @@
+#!/bin/bash
+# Copyright (C) 1989-2023 PC2 Development Team: John Clevenger, Douglas Lane, Samir Ashoo, and Troy Boudreau.
+#  
+# File:    pc2sandbox.sh 
+# Purpose: a sandbox for pc2 using Linux CGroups v2.
+# Input arguments:
+#  $1: memory limit in MB
+#  $2: time limit in seconds
+#  $3: command to be executed
+#  $4... : command arguments
+# 
+# Author: John Buck, based on earlier versions by John Clevenger and Doug Lane
+
+# FAIL_RETCODE_BASE is 128 + 64 + xx
+# 128 = system error, like signal
+# 64 = biggest used signal
+FAIL_RETCODE_BASE=192
+FAIL_EXIT_CODE=$((FAIL_RETCODE_BASE+43))
+FAIL_NO_ARGS_EXIT_CODE=$((FAIL_RETCODE_BASE+44))
+FAIL_INSUFFICIENT_ARGS_EXIT_CODE=$((FAIL_RETCODE_BASE+45))
+FAIL_INVALID_CGROUP_INSTALLATION=$((FAIL_RETCODE_BASE+46))
+FAIL_MISSING_CGROUP_CONTROLLERS_FILE=$((FAIL_RETCODE_BASE+47))
+FAIL_MISSING_CGROUP_SUBTREE_CONTROL_FILE=$((FAIL_RETCODE_BASE+48))
+FAIL_CPU_CONTROLLER_NOT_ENABLED=$((FAIL_RETCODE_BASE+49))
+FAIL_MEMORY_CONTROLLER_NOT_ENABLED=$((FAIL_RETCODE_BASE+50))
+FAIL_MEMORY_LIMIT_EXCEEDED=$((FAIL_RETCODE_BASE+51))
+FAIL_TIME_LIMIT_EXCEEDED=$((FAIL_RETCODE_BASE+52))
+FAIL_WALL_TIME_LIMIT_EXCEEDED=$((FAIL_RETCODE_BASE+53))
+
+# Process ID of submission
+submissionpid=""
+
+CGROUP_PATH=/sys/fs/cgroup
+PC2_CGROUP_PATH=$CGROUP_PATH/pc2
+# TODO: should have unique runbox in case multiple instances are executing in separate judges
+# For now, just use userid
+PC2_SANDBOX_CGROUP_PATH=$PC2_CGROUP_PATH/sandbox_$USER
+
+# control whether the script outputs debug/tracing info
+_DEBUG="on"   # change this to anything other than "on" to disable debug/trace output
+DEBUG_FILE=sandbox.log
+function DEBUG()
+{
+  [ "$_DEBUG" == "on" ] && $@ >> $DEBUG_FILE
+}
+
+# ------------------------------------------------------------
+
+usage()
+{
+  cat <<SAGE
+Usage: pc2_sandbox.sh memlimit timelimit command command_args
+
+memlimit, in MB
+timelimit, in seconds
+
+SAGE
+}
+
+# Function to handle getting killed by PC2's execute timer (basically, this
+# is wall-time exceeded which is execute time limit + 1 second
+HandleTerminateFromPC2()
+{
+	DEBUG echo "Received TERMINATE signal from PC2"
+	if test -n "$submissionpid"
+	then
+		DEBUG echo "Killing off submission process $submissionpid"
+		kill -9 "$submissionpid"
+	fi
+	DEBUG echo $0: Wall time exceeded - exiting with code $FAIL_WALL_TIME_LIMIT_EXCEEDED
+	exit $FAIL_WALL_TIME_LIMIT_EXCEEDED 
+}
+
+# ------------------------------------------------------------
+
+if [ "$#" -lt 1 ] ; then
+   echo $0: No command line arguments
+   exit $FAIL_NO_ARGS_EXIT_CODE
+fi 
+
+if [ "$#" -lt 3 ] ; then
+   echo $0: expected 3 or more arguments, found: $*
+   exit $FAIL_INSUFFICIENT_ARGS_EXIT_CODE
+fi 
+
+if [ "$1" = "-h" -o "$1" = "--help" ] ; then
+   usage
+   exit $FAIL_EXIT_CODE
+fi 
+
+MEMLIMIT=$1
+TIMELIMIT=$2
+COMMAND=$3
+shift
+shift
+shift
+
+#### Debugging - just set expected first 3 args to: 16MB 5seconds
+###MEMLIMIT=8
+###TIMELIMIT=2
+###COMMAND=$1
+###shift
```

Comment:
It looks to me like the above statements are intended to work like "remove the #'s at the beginning to enable this debugging facility".  However, it ALSO looks to me like this won't work, because it appears that the "command" argument has _already been shifted_ (by line 96, above).

Reply:
Yeah, should be removed; it was a "one or the other" type thing.  The comment above on 98 is wrong too.  All that should be removed.